File: LeetCode/Java Solutions/Reorganize String/reorganize.java
import java.util.HashMap;
import java.util.PriorityQueue;

class Solution {
    public String reorganizeString(String s) {
        // Step 1: Count occurrences of each character
        HashMap<Character, Integer> counts = new HashMap<>();
        for (char c : s.toCharArray()) {
            counts.put(c, counts.getOrDefault(c, 0) + 1);
        }

        // Step 2: Create a max-heap
        PriorityQueue<Character> maxHeap = new PriorityQueue<>((a, b) -> counts.get(b) - counts.get(a));
        maxHeap.addAll(counts.keySet());

        // If most frequent character is more than half of string, then impossible to rearrange
        if (counts.get(maxHeap.peek()) > (s.length() + 1) / 2) {
            return "";
        }

        // Step 3: Arrange characters by placing the most frequent characters first
        char[] arr = new char[s.length()];
        int index = 0;

        while (!maxHeap.isEmpty()) {
            char current = maxHeap.poll();
            int times = counts.get(current);
            for (int i = 0; i < times; i++) {
                // If the even indexes are exhausted, start from the first odd index
                if (index >= arr.length) {
                    index = 1;
                }
                arr[index] = current;
                index += 2;
            }
        }

        return new String(arr);
    }
}

File: LeetCode/Python Solutions/Reorganize String/reorganize.py
import heapq
from collections import Counter

class Solution:
    def reorganizeString(self, s: str) -> str:
        # Step 1: Count occurrence of each character
        counts = Counter(s)
        
        # Step 2: Create a max-heap
        max_heap = [(-count, char) for char, count in counts.items()]
        heapq.heapify(max_heap)
        
        # If most frequent character is more than half of string, then impossible to rearrange
        if -max_heap[0][0] > (len(s) + 1) // 2:
            return ""
        
        # Step 3: Arrange characters one by one using the heap
        prev_char = ''
        result = []
        
        while max_heap:
            count, char = heapq.heappop(max_heap)
            
            # If the current character is not same as previous, add to result
            if char != prev_char:
                result.append(char)
                if count < -1:
                    heapq.heappush(max_heap, (count + 1, char))
                prev_char = char
            else:
                # If we can't get another character, then return ""
                if not max_heap:
                    return ""
                
                next_count, next_char = heapq.heappop(max_heap)
                result.append(next_char)
                if next_count < -1:
                    heapq.heappush(max_heap, (next_count + 1, next_char))
                prev_char = next_char
                
                heapq.heappush(max_heap, (count, char))
        
        return "".join(result)
